test/test_replace.py

```diff
@@ -0,0 +1,225 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass, field
+
+import pytest
+
+from simple_parsing import replace, subgroups
+
+from .test_utils import TestSetup
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class A(TestSetup):
+    a: float = 0.0
+
+
+@dataclass
+class B(TestSetup):
+    b: str = "bar"
+    b_post_init: str = field(init=False)
+
+    def __post_init__(self):
+        self.b_post_init = self.b + "_post"
+
+
+@dataclass
+class AB(TestSetup):
+    integer_only_by_post_init: int = field(init=False)
+    integer_in_string: str = "1"
+    a_or_b: A | B = subgroups({"a": A, "b": B}, default="a")
+
+    def __post_init__(self):
+        self.integer_only_by_post_init = int(self.integer_in_string)
+
+
+@dataclass
+class C:
+    c: bool = False
+
+
+@dataclass
+class D:
+    d: int = 0
+
+
+@dataclass(frozen=True)
+class CD(TestSetup):
+    c_or_d: C | D = subgroups({"c": C, "d": D}, default="c")
+
+    other_arg: str = "bob"
+
+
+@dataclass
+class NestedDataclasses:
+    nested_arg: str = "nested_arg"
+    arg1: int = 0
+    arg2: float = 0.0
+
+
+@dataclass
+class OuterDataclass(TestSetup):
+    some_arg: str = "some_arg"
+    nested: NestedDataclasses = field(default_factory=NestedDataclasses)
+
+
+@pytest.mark.parametrize(
+    ("config", "changes"),
+    [
+        (A(a=2.0), {"a": 2.0}),
+        pytest.param(A(), {"b": 1.0}, marks=pytest.mark.xfail(reason="Unrecognized arguments")),
+        (B(b="test"), {"b": "test"}),
+        (AB(a_or_b=A()), {"a_or_b": "a"}),
+        (AB(a_or_b=B(b="foo")), {"__subgroups__@a_or_b": "b", "a_or_b": {"b": "foo"}}),
+        (AB(integer_in_string="2"), {"integer_in_string": "2"}),
+        pytest.param(
+            AB(integer_in_string="2"),
+            {"integer_in_string": "2", "integer_only_by_post_init": 2},
+            marks=pytest.mark.xfail(reason="Any field with init=False will raise ValueError"),
+        ),
+        (CD(), {}),
+        pytest.param(
+            "str_obj",
+            {},
+            marks=pytest.mark.xfail(reason="Raise TypeError if obj is not dataclass instances"),
+        ),
+        (
+            OuterDataclass(nested=NestedDataclasses(nested_arg="nested_arg_2", arg1=1, arg2=2.0)),
+            {"nested": {"nested_arg": "nested_arg_2", "arg1": 1, "arg2": 2.0}},
+        ),
+    ],
+)
+def test_replace_nested_dict(config: object, changes: dict):
+    config_replaced = replace(type(config)(), changes)
+    logger.info(config_replaced)
+    assert config == config_replaced
+    assert id(config) != id(config_replaced)
+
+
+@pytest.mark.parametrize(
+    ("config", "changes"),
+    [
+        (CD(c_or_d=D(d=1)), {"__subgroups__@c_or_d": "d", "c_or_d.d": 1}),
+        (CD(c_or_d=C(c=True)), {"__subgroups__@c_or_d": "c", "c_or_d.c": True}),
```

Comment:
Not sure that I understand the goal of these. Again, perhaps I'm wrong, feel free to tell me what you think about this one.
Is the idea to allow people to say "Switch the subgroup to 'd' and customize just this value with respect to the default" ?
If so, does this here look a lot better than the equivalent with a more traditional `replace` function?
```python
C_D_subgroups = {"c": C, "d": D}

@dataclass
class Config:
    c_or_d: C | D = subgroups(C_D_subgroups, default="c")

config = Config()
replaced_config = replace(config, c_or_d=C_D_subgroups["d"](d=1))
```

Reply:
Yeah, you are right. We can do both ways, Wouldn't that be more flexiable for users? I will add your suggestion to the test.